docs/proposal/kcl-checksum-verification-proposal.md

```diff
@@ -0,0 +1,208 @@
+- Modify the CheckSum function to read data from dependencies in the form of streams and calculate a consistent checksum based on the data streams:
+Existing Checksum Function:- https://github.com/kcl-lang/kpm/blob/2b1e0b91d8488da804158728063308ca271a3350/pkg/utils/utils.go#L32-L70
+
+We can take some motivation from Go's Hash function -
+https://github.com/golang/go/blob/f428c7b729d3d9b37ed4dacddcd7ff88f4213f70/src/cmd/vendor/golang.org/x/mod/sumdb/dirhash/hash.go#L44-L65
+
+Check for Go Implementations of generating hash while downloading the package in form of stream:- https://github.com/golang/go/blob/master/src/cmd/vendor/golang.org/x/mod/sumdb/dirhash/hash.go
+
+Fetching Example -
+https://github.com/golang/go/blob/f428c7b729d3d9b37ed4dacddcd7ff88f4213f70/src/cmd/go/internal/modfetch/fetch.go#L647-L652
+
+3. Add custom logic in `Download` function to generate the checksum from data streams using the updated `HashDir` function:
+
+https://github.com/kcl-lang/kpm/blob/2b1e0b91d8488da804158728063308ca271a3350/pkg/client/client.go#L1092-L1267
+
+4. `Download` function will load the package from `kcl.mod.lock` using `LoadPkgFromPath` 
+https://github.com/kcl-lang/kpm/blob/2b1e0b91d8488da804158728063308ca271a3350/pkg/client/client.go#L143-L187
+
+5. Compare the actual checksum with the current checksum using `check` function that will compare the checksums and, if matched, store the verified hash in the `kcl.mod.lock` file and load the package:
+
+https://github.com/kcl-lang/kpm/blob/2b1e0b91d8488da804158728063308ca271a3350/pkg/package/package.go#L238-L250
+
+6. If the checksums match, update the mod and lock file using `UpdateModAndLockFile`; otherwise, log an error message to the user to avoid downloading a malicious package.
+
+Additionally, we can extend it to verify this checksum every time for all existing or upcoming packages/dependencies if there is any modification in the `kcl.mod` and `kcl.mod.lock` files. Currently, KPM shows the sum only for the first time and does not give errors if the sum is changed and a download event is triggered. This enhancement would bring KPM in line with other package managers like Go, npm, Cargo, etc., which trigger reverification.
+
+## For the existing kpm dependencies, we need a solution that does not break compatibility for starting this feature
+We can have two methods to ensure backward compatibility:
+
+1. Checksum Registry/Database:
+
+We can take some motivation from npm, where we create a `registry` that contains metadata for all the packages and dependencies known to KPM. This is similar to the Go checksum database, which has `checksum database` for all the modules known to Go.
+If we are not able to find the checksum entry in the lockfile, we can retrieve it from the registry or `checksum database`. We can also add a flag, similar to Go, allowing users to skip checking the registry or checksum database and download the package even if it might be corrupted.
+
+2. FLAG_NO_SUM_CHECK:
```

Comment:
For this section, it is important to clarify when checksums verification are requested in kpm, and what impact they may have on the current kpm process.

Combined with the current `kcl run`, `kcl mod update`, `kcl mod push`, `kcl mod pull` , sorts out their current work flow and the effective position of checksums verification.

You can use a simple illustration as the example to show the workflow for kpm：

```
kcl run -> step 1 -> step 2 -> .......
```

And checksum verification will add as :

```
              verify sum here 
                       |
kcl run -> step 1 -> step 2 -> ......
               | 
      require sum here 
```

Reply:
Hey @zong-zhe, do we need checksum verification in `kcl run`? I navigated through the workflow of `kcl run` but could not find any place where it downloads or fetches modules from Git or OCI registry as part of its operation. Instead, it only fetches information from the local storage path of external packages.

Reply:
Hi @NishantBansal2003 😄 

kpm downloads OCI and GIT related content via Downloader: https://github.com/kcl-lang/kpm/blob/main/pkg/downloader/downloader.go

Reply:
Thank you for the clarification. I apologize if my previous message was unclear. What I meant to convey is that when we trigger the `kcl run` command in the CLI, the functions that are executed do not seem to involve downloading content from Git or an OCI registry. Instead, they appear to only access information from the local storage path of external packages.

Additionally, I wanted to clarify whether the workflow example you provided was just an illustrative example or if we actually need to explore the possibility of incorporating checksum verification during the kcl run command. Should I investigate how checksum verification might be integrated into the kcl run process, or was that meant as a general example?

Could you please confirm if I’m understanding both aspects correctly?